core/src/main/scala/kafka/log/Log.scala

```diff
@@ -393,27 +396,42 @@ class Log(@volatile var dir: File,
           case _: NoSuchFileException =>
             error(s"Could not find offset index file corresponding to log file ${segment.log.file.getAbsolutePath}, " +
               "recovering segment and rebuilding index files...")
-            recoverSegment(segment)
+            doSegmentRecovery = true
           case e: CorruptIndexException =>
             warn(s"Found a corrupted index file corresponding to log file ${segment.log.file.getAbsolutePath} due " +
               s"to ${e.getMessage}}, recovering segment and rebuilding index files...")
-            recoverSegment(segment)
+            doSegmentRecovery = true
         }
+
+        if (doSegmentRecovery) {
+          try recoverSegment(segment)
+          catch {
+            case e: IndexOffsetOverflowException =>
+              // Split the segment
+              Log.splitSegmentOnOffsetOverflow(this, segment)
+              // Close all segments that were opened, including the current segment we were working on. Throw the exception
+              // up for callers to retry if needed.
+              segment.close()
+              close()
```

Comment:
This requires that `Log` remains usable after a call to `close()`. Additionally, this method is called before the log has been fully initialized, which seems unsafe. It may work today, but it feels a little brittle to assume it going forward. Maybe it would be better to close the segments directly and remove them from `segments`?